### ISSUE TITLE
copy / paste puts objects over the top of copied objects

### ISSUE THREAD

**Issue by megrimm** [pure-data/pure-data#155]
this has been around a while and thought i would submit as issue.

when copy/paste selected objects, objects are pasted over the top of the original copied objects. 

1) copy / paste could paste offset like duplicate does
2) copy / paste could paste based on mouse location so as to have different functionality than duplicate.

there is much discussion on the list/forum about this and i think it makes it difficult for students, beginners when that are used to just copy/paste to paste a large selection of patch, try to more section and just get one object move and then fast lose selection.

the issue is for in same doc NOT when copy paste from one doc to another.

maybe this is just up for discussion. im taking to @porres on facebook about it now.

**Comment by danomatika** [comment]
IMO tThere doesn't need to be much discussion because this isn't useful for *anyone*.

**Comment by danomatika** [comment]
I made a quick try for this last night by displacing the pasted objects in relation to the mouse position. The pasting works fine but the positioning isn't right:

In g_editor.c canvas_paste:
~~~
        t_selection *y;
        int x1, y1, x2, y2;
        canvas_setundo(x, canvas_undo_paste, canvas_undo_set_paste(x),
            "paste");
        canvas_dopaste(x, EDITOR->copy_binbuf);
        for (y = x->gl_editor->e_selection; y; y = y->sel_next)
        {
            /* displace toward mouse pos */
            gobj_getrect(y->sel_what, x, &x1, &y1, &x2, &y2);
            gobj_displace(y->sel_what, x,
                x1 - x->gl_editor->e_xwas,
                y2 - x->gl_editor->e_ywas);
        }
        canvas_dirty(x, 1);
~~~

**Comment by porres** [comment]
Would we loose being able to paste to the pixel location? Cause I sometimes rely on that, specially when working with documentation. Could we have a shortcut to control it in any case?

**Comment by danomatika** [comment]
>Could we have a shortcut to control it in any case?

One option could be to have two behaviors:

* MOD+V: paste to pixel position
* MOD+Shift+V: paste to mouse position

This would be similar to how text is handled in most macOS app where you can paste with the copied styling (CMD+V) or paste and patch the current style (MOD+Alt+Shift+V).

**Comment by umlaeute** [comment]
i partially like the "paste to pixel position", as it allows to keep alignment of the pasted objects (they are within the 10 pixel steps i get with <kbd>Shift</kbd>-<kbd>↑</kbd>.

Obviously one can get the same behaviour with a default pasting displacement that is a multiple of the big movement steps (10 pix)), so I'd favour this (instead of a default "paste to mouse position")

i like the *optional* "paste to mouse pos" (keeping it an option), but would like to point out that the <kbd>Shift</kbd>-<kbd>Control</kbd>-<kbd>V</kbd> (which is what  <kbd>MOD</kbd>-<kbd>Shift</kbd>-<kbd>V</kbd>translates to on linux and w32) is already taken as `[vsl]` shortcut.

**Comment by danomatika** [comment]
You're right. [vsl] is created with MOD+Shift+V on macOS. As of my GUI updates from a year ago or so, all of the keybindings are the same across the platforms, except for which key is MOD.

**Comment by umlaeute** [comment]
i think this has been resolved with 70868534dc6e9cacc0c7e89a95b1d9bf6b55bd8a
